### DISCUSSION TITLE
Add doping profile as LayerLevel

### DISCUSSION THREAD

**Discussion by thomasdorch** [gdsfactory/gdsfactory#524] [Ideas]
Hi all,

I think it would be nice to have process information in the layer stacks, like whether the LayerLevel is a grow, etch, implant, or background layer, as this would allow for doping profiles to be added to the stack and would remove the need to add a cladding material when setting up the simulation. Additionally, it would make it easier to import a layerbuilder process file from Lumerical, as this is similar to how it's set up there.

For instance, this is how I would envision a stack for a deep etch PN junction phase shifter (for compactness, just P and N regions):
```
LayerStack(
        layers=dict(
            core=LayerLevel(
                layer=LAYER.WG,
                thickness=thickness_wg,
                zmin=0.0,
                material="si",
                process="grow"
            ),
            clad=LayerLevel(
                layer=LAYER.WGCLAD,
                zmin=0.0,
                material="sio2",
                thickness=thickness_clad,
                process="background"
            ),
            p=LayerLevel(
                layer=LAYER.P,
                thickness=thickness_wg,
                zmin=0.0,
                material="si_p",
                process="implant"
            ),
            n=LayerLevel(
                layer=LAYER.N,
                thickness=thickness_wg,
                zmin=0.0,
                material="si_n",
                process="implant"
            ),
            etch90=LayerLevel(
                layer=LAYER.SLAB90,
                thickness=wg_thickness - 90e-3,
                zmin=90e-3,
                material="sio2",
                process="etch"
            ),
      )
)
```
Then, under the hood, the geometry setup might look something like:
- add a prism for the growth layer using core material (across the whole width of the simulation domain),
- add prisms for implant layers using implant material,
- add prisms for each etch using etch material,
- add prisms for full etch using clad/background material

Are there any thoughts on doing something like this?

**Comment by joamatab** [comment]
I like it 


Dphox had something like that
@sunilkpai

**Reply by joamatab** [reply]
We can also add refractive index data for each material 

How about allowing both an int, complex or callable ?

**Reply by simbilod** [reply]
If we are adding metadata, I would also allow adding information for the electrical layers like dopant concentration and type (for doping layers), and resistivity (for metals and vias)

**Reply by simbilod** [reply]
Maybe all of this could be a "simulation_info" dict that users can extend themselves? With some good default for standard simulator use cases

**Reply by thomasdorch** [reply]
I really like the way Dphox implements a `Foundry` and passes in the layer stack to it.. 

What about only adding a process info string (grow, etch, doping, background) that dictates how the geometry is created, and using a separate Material BaseModel (or dict) to store the relevant info about the material (doping concentration, refractive index, resistivity, etc.), then collecting them into a Materials dict? That way the user can add whatever extra info they want to a material, provided that it at least has a `refractive_index` attribute.

**Reply by thomasdorch** [reply]
Using BaseModels for materials would also be nice in the case of dopants specifically because this implementation lends itself to having an inheritance like this:
```
class Si(BaseModel):
    refractive_index
    
    def get_index():
        ...

class n_Si(Si):
    index_perturbation: callable
    doping_concentration: float
    
    def get_index(self): 
        return self.index_perturbation(super().get_index(), self.doping_concentration)
``` 
so the user is responsible for setting up a material model but the BaseModel helps validate the attributes that are required.

**Comment by joamatab** [comment]
I like the idea of simulation_info

https://github.com/gdsfactory/gdsfactory/pull/526

once we add Tcad plugins for devsim, Sentaurus, charon or Lumerical device, we can polish it further

another option is having a layerType field 

```
class LayerType(BaseModel):
  pass

class LayerDoped(LayerType):
  doping_concentration: float =1e-9
  doping_type: str = 'p'

class LayerMetal(LayerType):
  resistivity:float

````

what do you think?

@flaport 
@momchil-flex 
@mdecea

**Comment by flaport** [comment]
This is how I have extruded a GDS recently, it's pretty similar actually

```python
extrusion_rules = {
    LAYERS.WGCLAD: [ # multiple rules per layer
        GdsExtrusionRule(
            material=silicon, # the material is an object with its own properties, like process info
            h_min=0.0,
            h_max=0.0 + t_slab,
            mesh_order=1, # an extra mesh order is always useful
        ),
        GdsExtrusionRule(
            material=silicon_oxide,
            h_min=0.0,
            h_max=0.0 + t_slab + t_ox,
            mesh_order=2,
        ),
    ],
    LAYERS.WGCORE: [
        GdsExtrusionRule(
            material=silicon,
            h_min=0.0,
            h_max=0.0 + t_soi,
            mesh_order=1,
        ),
        GdsExtrusionRule(
            material=silicon_oxide,
            h_min=0.0,
            h_max=0.0 + t_soi + t_ox,
            buffer=t_ox, # grow-shrink operation
            mesh_order=2,
        ),
    ],
}
```

This is basically a simple rule to extrude a rib waveguide.

Finally, I think having a mesh order can still be pretty useful for extrusions.

**Comment by joamatab** [comment]
We added layer type to layerLevel last week

I hope you find it useful
